src/java.base/share/classes/javax/crypto/EncryptedPrivateKeyInfo.java

```diff
@@ -492,68 +501,110 @@ private static EncryptedPrivateKeyInfo encryptKeyImpl(PrivateKey key,
-     * decrypting the encrypted data, or producing the key object.
-     * @throws NullPointerException if {@code decryptKey} is null
+     * @return a {@code PrivateKey} extracted from the encrypted data
+     * @throws GeneralSecurityException if an error occurs parsing,
+     * decrypting the data, or producing the key object.
+     * @throws NullPointerException if {@code decryptKey} is {@code null}
      *
      * @since 25
      */
     @PreviewFeature(feature = PreviewFeature.Feature.PEM_API)
     public PrivateKey getKey(Key decryptKey, Provider provider)
         throws GeneralSecurityException {
         Objects.requireNonNull(decryptKey,"decryptKey cannot be null.");
-        PKCS8EncodedKeySpec p = getKeySpecImpl(decryptKey, provider);
-        try {
-            if (provider == null) {
-                return KeyFactory.getInstance(
-                    KeyUtil.getAlgorithm(p.getEncoded())).
-                    generatePrivate(p);
-            }
-            return KeyFactory.getInstance(KeyUtil.getAlgorithm(p.getEncoded()),
-                provider).generatePrivate(p);
-        } catch (IOException e) {
-            throw new GeneralSecurityException(e);
-        }
+        return PKCS8Key.parseKey(decryptData(decryptKey, provider), null);
+    }
+
+    /**
+     * Extracts and returns a {@code KeyPair} object from the encrypted data.
+     * The data must contain both a public and private key.
+     *
+     * @param password the password used in the PBE encryption.  This array
+     *                 will be cloned before being used.
+     * @return a {@code KeyPair} extracted from the encrypted data
+     * @throws GeneralSecurityException if an error occurs while parsing,
+     * decrypting the data, or producing the {@code KeyPair} object.
+     * @throws NullPointerException if {@code password} is {@code null}
+     *
+     * @since 26
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.PEM_API)
+    public KeyPair getKeyPair(char[] password) throws GeneralSecurityException {
+        Objects.requireNonNull(password, "password cannot be null");
+
+        DEREncodable d = Pem.toDEREncodable(
+            Pem.decryptEncoding(encoded, password), true, null);
+        return switch (d) {
+            case KeyPair kp -> kp;
+            case PrivateKey ignored -> throw new InvalidKeyException(
+                "This encoding do not contain a public key.");
+            default -> throw new InvalidKeyException(
+                "Invalid class returned " + d.getClass().getName());
+        };
+    }
+
+    /**
+     * Extracts and returns a {@code KeyPair} object from the encrypted data.
+     * The data must contain both a public and private key.
+     *
+     * @param decryptKey the decryption key and cannot be {@code null}
+     * @param provider the {@code Provider} used for Cipher decryption and
+     *                 key generation. A {@code null} value will
```

Comment:
Why do you allow null? This is inconsistent with the other methods that take Provider arguments. I think consistency is important.